### PR TITLE
373 sqlhistory snap to bronze

### LINKED ISSUE #134: [EPIC] - Publish DBSQL Gold Entities
Originally, SQL Analytics Endpoints were identified as specialized clusters but they have since been modified into an entirely separate entity. As such, at this time, no SQL analytics data is being captured in the clusters_snapshot table but further investigation is necessary to determine how the audit logs have changed.

Targeting this resolution for next release 0.4.2

## Scope and Steps

1.  Get the dataframe from API . This will be defined in Workspace.scala. It will be similar to that of clusers_snapshot_bronze and jobs_snapshot_bronze this will act as a lookup for missing audit data in Silver
2. Define a endPointsTarget.  Just make sure it's defined as a Bronze target
3. Create a module inside Bronze.   Make sure it has an appropriate module ID and name. The schema scrubber implementation should NOT be INSIDE BRONZE -- but rather a function defined inside the trait, BronzeTransforms -- as needed
4. You create a process that implements ETL Definition inside Bronze.scala . You create a process that implements ETL Definition inside Bronze.scala -- this is what will actually call the functions in the BronzeTransformations that perform the schema cleansing, etc
5. Create a pipeline to execute the module and load the Target.  Bronze() is a pipeline so all you need to do is check the Overwatch scope to see if sqlAnalytics is present -- if it is, execute the process defined in Step 4, otherwise skip
which leads to step 0-ish which is likely add sqlAnalytics scope to the Overwatch Scope enumeration object and validate any requisite dependencies (there may not be any in bronze)

### LINKED ISSUE #216: Include SQL Queries history into collected data
[SQL History API](https://docs.databricks.com/sql/api/query-history.html) allows to pull the data about executed queries, together with information about execution time, status, etc. It would be beneficial for some reports

### LINKED ISSUE #373: Add Sql Query History Bronze Table
Use ApiV2 to land the SQL Query History

Related to #216 and #134

### PR DESCRIPTION
**Pull request by aman-db**
closes #373 
closes #216 
partially closes #134 
contains code for - 
* Bronze , silver and gold sqlHistory

### CODE REVIEW

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Silver.scala [hunk 1]
```diff
@@ -301,6 +303,13 @@ class Silver(_workspace: Workspace, _database: Database, _config: Config)
     append(SilverTargets.notebookStatusTarget)
   )
 
+  lazy private[overwatch] val sqlHistoryModule = Module(2020, "Silver_SqlHistory", this, Array(1016))
+  lazy private val appendSqlHistoryProcess = ETLDefinition(
+    BronzeTargets.sqlHistorySnapshotTarget.asIncrementalDF(sqlHistoryModule, SilverTargets.sqlHistoryTarget.incrementalColumns),
+    Seq(sqlHistoryTransform()),
+    append(SilverTargets.sqlHistoryTarget)
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Silver.scala:310]: This silver target is a WriteMode.merge but you aren't selecting any overlapping data. What happens when a query starts during run 1 and finishes in run 2 -- does it just never get its closure? Please select 2 lag days on this input and validate your keys and uniqueness across those keys to enable overlapping merges and close open queries across OW runs.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Gold.scala [hunk 2]
```diff
@@ -218,6 +222,12 @@ class Gold(_workspace: Workspace, _database: Database, _config: Config)
     append(GoldTargets.sparkExecutorTarget)
   )
 
+  lazy private[overwatch] val sqlHistoryModule = Module(3020, "Gold_SqlHistory", this, Array(2020), 6.0)
+    .withSparkOverrides(sparkBaseSparkOverrides)
+  lazy private val appendSqlHistoryProcess = ETLDefinition(
+    SilverTargets.sqlHistoryTarget.asIncrementalDF(sqlHistoryModule, GoldTargets.sqlHistoryTarget.incrementalColumns),
+    append(GoldTargets.sqlHistoryTarget)
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Gold.scala:229]: This gold target is a WriteMode.merge but you aren't selecting any overlapping data. What happens when a query starts during run 1 and finishes in run 2 -- does it just never get its closure? Please select 2 lag days on this input and validate your keys and uniqueness across those keys to enable overlapping merges and close open queries across OW runs.

**Review by GeekSheikh — CHANGES_REQUESTED** [review]
Very close -- thanks. Should be quick changes.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Bronze.scala [hunk 3]
```diff
@@ -199,13 +207,15 @@ class Bronze(_workspace: Workspace, _database: Database, _config: Config)
   }
 
   private def executeModules(): Unit = {
+    config.overwatchScope.foreach(x=>println(x))
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Bronze.scala:210]: yes that was just for debug. removed it.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Bronze.scala [hunk 4]
```diff
@@ -199,13 +207,15 @@ class Bronze(_workspace: Workspace, _database: Database, _config: Config)
...
     config.overwatchScope.foreach {
       case OverwatchScope.audit => auditLogsModule.execute(appendAuditLogsProcess)
       case OverwatchScope.clusters => clustersSnapshotModule.execute(appendClustersAPIProcess)
       case OverwatchScope.clusterEvents => clusterEventLogsModule.execute(appendClusterEventLogsProcess)
       case OverwatchScope.jobs => jobsSnapshotModule.execute(appendJobsProcess)
       case OverwatchScope.pools => poolsSnapshotModule.execute(appendPoolsProcess)
       case OverwatchScope.sparkEvents => sparkEventLogsModule.execute(appendSparkEventLogsProcess)
+      case OverwatchScope.sqlHistory => sqlHistorySnapshotModule.execute(appendSqlHistoryProcess)
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Bronze.scala:218]: got it. Moved

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Gold.scala [hunk 5]
```diff
@@ -273,6 +284,10 @@ class Gold(_workspace: Workspace, _database: Database, _config: Config)
       case OverwatchScope.sparkEvents => {
         processSparkEvents()
       }
+      case OverwatchScope.sqlHistory => {
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Gold.scala:287]: Done.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Gold.scala [hunk 6]
```diff
@@ -218,6 +222,12 @@ class Gold(_workspace: Workspace, _database: Database, _config: Config)
     append(GoldTargets.sparkExecutorTarget)
   )
 
+  lazy private[overwatch] val sqlHistoryModule = Module(3020, "Gold_SqlHistory", this, Array(2020), 6.0)
```
[the remark this answers is not in the document]
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Gold.scala:225]: We have used the module 2020 for Silver so thought of using 3020 for gold module. Changed it to 3017.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Silver.scala (hunk 1 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Silver.scala:310]: added additional lag of 2 days

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Gold.scala (hunk 2 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/Gold.scala:229]: added additional lag of 2 days

**Review by GeekSheikh — APPROVED** [review]

**Pull request merged**